lib/src/solver/solve_suggestions.dart

```diff
@@ -0,0 +1,214 @@
+              suggestion =
+                  '* Consider downgrading your constraint on $name: $topLevelProgram pub add $addDescription';
+            } else {
+              priority = 2;
+              suggestion =
+                  '* Try upgrading your constraint on $name: $topLevelProgram pub add $addDescription';
+            }
+          }
+        }
+
+        return _ResolutionSuggestion(suggestion, priority: priority);
+      }
+    }
+    return null;
+  }
+
+  Future<_ResolutionSuggestion?> suggestUnlockingAll(
+      {required bool stripLowerBound}) async {
+    final originalPubspec = entrypoint.root.pubspec;
+    final relaxedPubspec =
+        stripVersionBounds(originalPubspec, stripLowerBound: stripLowerBound);
+
+    final result = await tryResolve(
+        type, cache, Package.inMemory(relaxedPubspec),
+        lockFile: entrypoint.lockFile, unlock: unlock);
+    if (result != null) {
+      final updatedPackageVersions = <PackageId>[];
+      for (final id in result.packages) {
+        final originalConstraint = (originalPubspec.dependencies[id.name] ??
+                originalPubspec.devDependencies[id.name])
+            ?.constraint;
+        if (originalConstraint != null) {
+          updatedPackageVersions.add(id);
+        }
+      }
+      if (stripLowerBound && updatedPackageVersions.length > 5) {
+        // Too complex, don't suggest.
+        return null;
+      }
+      if (stripLowerBound) {
+        updatedPackageVersions.sort((a, b) => a.name.compareTo(b.name));
+        final formattedConstraints = updatedPackageVersions
+            .map((e) => packageAddDescription(entrypoint, e))
+            .join(' ');
+        return _ResolutionSuggestion(
+            '* Try updating the following constraints: $topLevelProgram pub add $formattedConstraints',
+            priority: 4);
+      } else {
+        return _ResolutionSuggestion(
+            '* Try an upgrade of your constraints: $topLevelProgram pub upgrade --major-versions',
+            priority: 4);
+      }
+    }
+    return null;
+  }
+
+  final visited = <String>{};
+  final stopwatch = Stopwatch()..start();
+  final suggestions = <_ResolutionSuggestion>[];
+  void addSuggestion(_ResolutionSuggestion? suggestion) {
+    if (suggestion != null) suggestions.add(suggestion);
+  }
+
+  for (final externalIncompatibility
+      in incompatibility.externalIncompatibilities) {
+    if (stopwatch.elapsed > Duration(seconds: 3)) {
+      // Never spend more than 3 seconds computing suggestions.
+      break;
+    }
+    final cause = externalIncompatibility.cause;
+    if (cause is SdkCause) {
+      addSuggestion(await suggestSdkUpdate(cause));
+    } else {
+      for (final term in externalIncompatibility.terms) {
+        final name = term.package.name;
+
+        if (!visited.add(name)) {
+          continue;
+        }
+        addSuggestion(await suggestSinglePackageUpdate(name));
+      }
+    }
+  }
+  if (suggestions.isEmpty) {
```

Comment:
The reason we don't want to suggest `upgrade --major-versions` is because we have a smaller suggestion, right?

Like if we look at the suggestions, all suggestions are:
 * A list of package names for which the constraints are tweaked
 * For each package where the constraints are tweaked:
    * Did we upgrade / downgrade?
    * Did we jump a major version or not?
    * How many versions did we jump?

Could we describe the size of a suggestion somehow? Then make the logic on whether include a suggestion more generic.

I think the space of all suggestions can be considered a partially-ordered-set, but if we want to do that we might have to think a bit :rofl:  maybe it's overkill.

Reply:
Yeah... that might make sense... still I'd like not having to compute the --major versions if we found an upgrade to a single package, however that would be required to do the compare...